### ISSUE TITLE
checkbox two-way binding issue

### ISSUE THREAD

**Issue by Calx-Takeno** [ractivejs/ractive#2833]
## Reproduction:
you can visit this URL
https://jsfiddle.net/calx/47fosorf/2/

## Code:none

<!--
Hello!
we find an issue above 0.7.3 about check two-way binding.

if you have fixed this , please tell me ,this is my email address 232613143@qq.com.

thank you for your efforts on git.

**Comment by Calx-Takeno** [comment]
when we observe the value change, it will get the same value between oldValue and newValue according to the document said.

**Comment by ghost** [comment]
Works on 0.6 https://jsfiddle.net/47fosorf/4/

**Comment by evs-chris** [comment]
There's a question here as to the correct behavior of the binding. You're observing an array, say `var arr = []` and you push an element onto it e.g. `arr.push(1)`. Now, the array has changed, so the observer will fire, but the old value and the new value are still the same array `arr`. Only some part of the array has changed and not the array itself.

The question is, should the checkbox binding completely replace the array when it updates, so you get a new array, or should it keep the existing one. I personally lean toward keeping the existing one, as a checkbox name binding feels like it should be the equivalent of a series of push/splice operations to match what happens to the checkboxes involved.

Any thoughts?

**Comment by sarendipitee** [comment]
>  I personally lean toward keeping the existing one, as a checkbox name binding feels like it should be the equivalent of a series of push/splice operations to match what happens to the checkboxes involved

Yep that makes most sense to me. Couldn't you still use `arr.*` anyways if you wanted to observe only push/slice ops?


Sidenote: The fact that this works at all is amazing. I think it's super convoluted to be storing multiple two-way bound checkboxes into a single array so am fine with undefined behavior there.

**Comment by ghost** [comment]
I was having doubts earlier, but it does seem related to https://github.com/ractivejs/ractive/issues/2639.

**Comment by evs-chris** [comment]
There is also the undocumented `observeList` function that takes a keypath, a callback, and options, and it will fire with an object that has inserted and deleted lists and a starting index. It doesn't really handle merges very well though, as they don't necessarily cluster around a starting index. I think we should probably rename it to `observeArray`, add in the oldIndex -> newIndex map for handling merges, and document it. It was added as somewhat experimental, but since the policy is now to add most new features as experimental pending feedback, I think this should be exposed more prominently.

https://github.com/ractivejs/ractive/blob/dev/src/Ractive/prototype/observeList.js
https://github.com/ractivejs/ractive/blob/dev/test/browser-tests/methods/observeList.js

**Comment by evs-chris** [comment]
_Or_ maybe we should move it to an option of `observe`? 
```js
ractive.observe(
  'some.list',
  changes => /* do things with changes.inserted and changes.deleted and perhaps a currently undefined changes.changes? */,
  { init: false, array: true }
);
```

**Comment by ghost** [comment]
How about a `clone` option for `observe` that when `true`:

- Shallow-copies values before and after mutation as `oldVal` and `newVal`, respectively.
- For object types which can't be cloned/doesn't make sense to do, don't clone. Use same object.
- If for some reason the observer should fire and no change done, don't clone. Use same object.
- For primitives, doesn't matter. 😄 

When `false` (default):

- Acts like the old observer system. Use same value regardless of type.

The default is the existing observer system because it makes more sense (mutating an object is the same object, nothing changed). Developer has to opt-in to the new setting to avoid breakage. The values are the full values, not just the diff since the use case is probably to cache old values after changes happen.

**Comment by evs-chris** [comment]
I suggested something similar when this came up before with objects, but we decided that that would probably be a bit too in depth to do usefully as a builtin. I think a general hook of some sort would be more appropriate so that the user can handle it however they need to. I was thinking an option for the observer would be a good way to let the user provide an old value for the next call, but looking at it, the return value of the callback isn't used, so it wouldn't be much effort at all to use a special return value for the old value. So

```js
ractive.observe('foo.bar', v => ..., { init: false, cache(v) { return _.cloneDeep(v); } });
//or
ractive.observe('foo.bar', v => { ...; return { cache: _.cloneDeep(v) }; }, { init: false });
```

I like also like `previous`, `cachePrevious`, and `nextOld` as names. `nextPrevious` is also interesting for oxymoronic reasons.
